nnmnkwii/datasets/jvs.py

```diff
@@ -0,0 +1,221 @@
+        self.labelmap = labelmap
+        self.labels = None
+        self.max_files = max_files
+        # TODO: currently there are many problem in falset, so we do not support falset
+        self._folders = ["parallel100", "nonpara30", "whisper10"]
+        self._textfilename = "transcripts_utf8.txt"
+        self._wavfoldername = "wav24kHz16bit"
+        self.speaker_info = _parse_speaker_info(data_root)
+        self._nonpara_no_wav = self._validate()
+
+    def _validate(self, strict=False):
+        # for each speaker, validate the data
+        # TODO: check lab file
+        nonpara_no_wav = {}
+        for _, speaker in enumerate(self.speakers):
+            # 1. check folder
+            speaker_folder = join(self.data_root, speaker)
+            folders = listdir(speaker_folder)
+            nonpara_no_wav[speaker] = {}
+            for folder in self._folders:
+                assert folder in folders,\
+                    "Can not find {} for {} in its directory {}"\
+                    .format(folder, speaker, speaker_folder)
+            # 2. check wav and txt file,
+            for folder in self._folders:
+                speaker_textfile = join(speaker_folder, folder, self._textfilename)
+                assert isfile(speaker_textfile), "File {} doesn't exist".format(speaker_textfile)
+                speaker_wavfolder = join(speaker_folder, folder, self._wavfoldername)
+                assert isdir(speaker_wavfolder), "Directory {} doesn't exist".format(speaker_wavfolder)
+                # nonpara
+                if folder == "nonpara30":
+                    with open(speaker_textfile, encoding="utf8") as file_:
+                        txtfiles = [line.strip().split(":")[0] for line in file_]
+                    wavfiles = listdir(speaker_wavfolder)
+                    for file_ in txtfiles:
+                        if file_ + ".wav" not in wavfiles:
+                            nonpara_no_wav[speaker][file_] = '_'
+                    assert (len(txtfiles) - len(nonpara_no_wav[speaker])) == len(wavfiles)
+                # strict mode
+                if strict:
+                    with open(speaker_textfile, encoding='utf8') as file_:
+                        txtlines = [line.strip().split(':')[0] for line in file_]
+                    wavlines = [basename(line) for line in listdir(speaker_wavfolder)]
+                    for line in txtlines:
+                        if line + '.wav' not in wavlines:
+                            print(speaker, speaker_wavfolder, line)
+        return nonpara_no_wav
+
+    def collect_files(self, is_wav, nonpara=False, whisper=False):
+        paths, labels = [], []
+        global lost_wavfiles
+        max_files_per_speaker = self.max_files // len(self.speakers) if self.max_files else None
+        for idx, speaker in enumerate(self.speakers):
+            speaker_folder = join(self.data_root, speaker)
+
+            def read_text(filepath, nonpara, para):
```

Comment:
Better to move function definitions to the outside of the class? unless there's any strong reason.

Reply:
The `read_text` function uses some parameters of the class method `collect_files`, thus it is difficult to more it out. `read_wavs` could be moved, but since this function has similar function with `read_text` so I think it would be better to remain inside the class.